### PR TITLE
Level 3

### PR DESCRIPTION
**Pull request by MaximKalinin**

### CODE REVIEW

File: src/Server.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        this._onMessage = this._onMessage.bind(this);
+    }
+    async listen(port?: number | ((port: number) => void), callback?: (port: number) => void)
+    {
+        this._port = typeof port === 'number' ? port : 8000;
+        this._socket = dgram.createSocket("udp4");
+        this._clients = {};
+
+        await new Promise(resolve => this._socket.bind(this._port, undefined, resolve));
+        this._socket.on('message', this._onMessage);
```
Comment (Pigotz) [inline src/Server.ts:20]: New line :see_no_evil:

File: src/Server.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        if (typeof callback === 'function') 
+            callback(this._port);
+        else if (typeof port === 'function')
+            port(this._port);
+    };
+    shutdown(callback?: () => void) 
+    {
+        if (!this._socket)
+            return;
+        this._socket.close(callback);
```
Comment (Pigotz) [inline src/Server.ts:31]: New line :see_no_evil:

File: src/Server.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        this._socket = undefined;
+        this._port = undefined;
+        this._clients = undefined;
+    }
+    private _onMessage(buffer: Buffer, rinfo: dgram.AddressInfo)
```
Comment (Pigotz) [inline src/Server.ts:36]: :snake:

File: src/Server.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                this._broadcast(message);
+                break;
+            case MessageType.LEAVE:
+                this._removeClient(message);
+                break;
+            default:
+                break;
+        }
+    }
+    private _addClient(message: IMessage, address: Address)
```
Comment (Pigotz) [inline src/Server.ts:58]: :snake:

File: src/Server.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                this._removeClient(message);
+                break;
+            default:
+                break;
+        }
+    }
+    private _addClient(message: IMessage, address: Address) 
+    {
+        const {source: {id, username}} = message;
+        if (this._clients[id])
```
Comment (Pigotz) [inline src/Server.ts:61]: New line :see_no_evil:

File: src/Server.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            default:
+                break;
+        }
+    }
+    private _addClient(message: IMessage, address: Address) 
+    {
+        const {source: {id, username}} = message;
+        if (this._clients[id]) 
+            return;
+        this._clients[id] = {id, username, address};
```
Comment (Pigotz) [inline src/Server.ts:63]: New line :see_no_evil:

File: src/Server.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    private _sendMessage(message: IMessage)
```
Comment (Pigotz) [inline src/Server.ts:65]: :snake:

File: src/Server.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        const {payload, destination} = message;
+        const client = this._clients[destination];
+        if (!client)
```
Comment (Pigotz) [inline src/Server.ts:69]: New line :see_no_evil:

File: src/Server.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            return;
+        const {address: {ip, port}} = client;
```
Comment (Pigotz) [inline src/Server.ts:71]: New line :see_no_evil:

File: src/Server.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    private _broadcast(message: IMessage) 
+    {
+        const {payload} = message;
+        Object.keys(this._clients).forEach(key => 
+        {
+            this._sendMessage({...message, destination: +key});
+        });
+    }
+    private _removeClient(message: IMessage)
```
Comment (Pigotz) [inline src/Server.ts:82]: :snake:

File: src/Server.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    private _sendMessage(message: IMessage)
+    {
+        const {payload, destination} = message;
+        const client = this._clients[destination];
+        if (!client)
+            return;
+        const {address: {ip, port}} = client;
+        this._socket.send(payload, port, ip);
+    }
+    private _broadcast(message: IMessage)
```
Comment (Pigotz) [inline src/Server.ts:74]: Missing return type

File: src/Server.ts (hunk 10 above)
Comment (Pigotz) [inline src/Server.ts:82]: Missing return type

File: src/Server.ts (hunk 7 above)
Comment (Pigotz) [inline src/Server.ts:65]: Missing return type

File: src/Server.ts (hunk 4 above)
Comment (Pigotz) [inline src/Server.ts:58]: Missing return type

File: src/Server.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        this._socket.close(callback);
+        this._socket = undefined;
+        this._port = undefined;
+        this._clients = undefined;
+    }
+    private _onMessage(buffer: Buffer, rinfo: dgram.AddressInfo) 
+    {
+        const {address, port} = rinfo;
+        const message: IMessage = JSON.parse(buffer.toString());
+        switch (message.type)
```
Comment (Pigotz) [inline src/Server.ts:40]: New line :see_no_evil:

File: src/Server.ts (hunk 3 above)
Comment (Pigotz) [inline src/Server.ts:36]: Missing return type

File: src/Server.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+        await new Promise(resolve => this._socket.bind(this._port, undefined, resolve));
+        this._socket.on('message', this._onMessage);
+
+        if (typeof callback === 'function') 
+            callback(this._port);
+        else if (typeof port === 'function')
+            port(this._port);
+    };
+    shutdown(callback?: () => void)
```
Comment (Pigotz) [inline src/Server.ts:27]: Missing return type

File: src/Server.ts [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+class Server 
+{
+    private _port: number;
+    private _socket: dgram.Socket;
+    private _clients: {[id: number] : IClient};
+    constructor() 
+    {
+        this._onMessage = this._onMessage.bind(this);
+    }
+    async listen(port?: number | ((port: number) => void), callback?: (port: number) => void)
```
Comment (Pigotz) [inline src/Server.ts:13]: Missing return type

**Review by Pigotz — CHANGES_REQUESTED** [review]

File: src/Client.ts [hunk 15]
```diff
@@ -0,0 +1,102 @@
+import * as dgram from 'dgram';
+import {Address, IMessage, MessageType} from './Interfaces';
+
+export default class
```
[the remark this answers is not in the document]
Comment (MaximKalinin) [reply src/Client.ts:4]: name of the class?

File: src/Client.ts [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+    private _server: Address;
+    constructor(id: number, username: string) 
+    {
+        this._id = id;
+        this._username = username;
+    }
+    async connect(server: Address = {ip: 'localhost', port: 8000}): Promise<Address>
+    {
+        const {port, ip} = server;
+        this._server = server;
```
[the remark this answers is not in the document]
Comment (MaximKalinin) [reply src/Client.ts:18]: after `const {port, ip} = server;`?

File: src/Client.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (Pigotz) [reply src/Client.ts]: I think `Client` could be a good name

File: src/Client.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (Pigotz) [reply src/Client.ts:21]: Yes, I didn't specify

**Review by Pigotz — CHANGES_REQUESTED** [review]
Now it's better but there are some comments still open. Also, in the last review I completely missed the fact that the `Client` class has no member access modifiers. Please, specify them :+1:

File: src/Client.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (MaximKalinin) [reply src/Client.ts]: Changed😉

File: src/Server.ts [hunk 17]
```diff
@@ -0,0 +1,89 @@
+import * as dgram from 'dgram';
+import {IClient, IMessage, MessageType, Address} from './Interfaces';
+
+class Server
```
[the remark this answers is not in the document]
Comment (MaximKalinin) [reply src/Server.ts:4]: Changed 😉

**Review by Pigotz — APPROVED** [review]
:+1:

**Pull request closed without merging**